doc/SONIC_Test_Ingress_Discards_HLD.md

```diff
@@ -0,0 +1,999 @@
+	- get_l2_pkt_drops
+	- get_l3_pkt_drops
+	- get_acl_pkt_drops
+
+#### SAI attributes
+```SAI_PORT_STAT_IF_IN_DISCARDS``` - number of L2 discards
+
+```SAI_ROUTER_INTERFACE_STAT_IN_ERROR_PACKETS``` - number of L3 discards
+
+#### General test flow
+##### Each test case will use the following port types:
+- VLAN (T0)
+- LAG (T0, T1-LAG)
+- Router (T1, T1-LAG, PTF32)
+
+##### Drop counters which are going to be checked
+It depends on test objective.
+
+Will be verified one of the following drop counters:
+- Drop counter for L2 discards
+- Drop counter for L3 discards
+- Drop counter for ACL discards
+
+##### Sent packet number:
+N=5
+
+##### step #1 - Disable VMs
+Before test suite run - disable control plane traffic generation. Use "testbed-cli.sh" script with "disconnect-vms" option.
+
+##### step #2 - Execute test scenario for all available port types depends on run topology (VLAN, LAG, Router)
+
+- Select PTF ports to send and sniff traffic
+- Clear counters on DUT. Use CLI command "sonic-clear counters"
+- Inject N packet via PTF port selected for TX
+- Check specific drop counter incremented on N (depends on test case)
+	- If counter was not incremented on N, test fails with expected message
+- Check other counters were not incremented on N based on sent packet type, sent port and expected drop reason (depends on test case)
+	- If counter was incremented on N, test fails with expected message
+- Check the packet was dropped by sniffing packet absence on PTF port selected for RX
+
+##### step #3 - Enable VMs
+Enable previously disabled VMs using "testbed-cli.sh" script with "connect-vms" option.
+
+#### Run test
+```
+py.test --inventory ../ansible/inventory --host-pattern [DEVICE] --module-path ../ansible/library/ --testbed [DEVICE-TOPO] --testbed_file ../ansible/testbed.csv --junit-xml=./target/test-reports/ --show-capture=no --log-cli-level debug -ra -vvvvv ingress_discard/test_ingress_discard.py
+```
+
+#### Test cases
+Each test case will be additionally validated by the loganalyzer utility.
+
+Each test case will run specific traffic to trigger specific discard reasone.
+
+Pytest fixture - "ptfadapter" is used to construct and send packets.
+
+After packet is sent using source port index, test framework waits during 5 seconds for specific packet did not appear, due to ingress packet drop, on one of the destination port indices.
+
+#### Test case #1
+##### Test objective
+
+Verify packet drops when SMAC and DMAC are equal
+
+Packet to trigger drop
+```
+###[ Ethernet ]###
+  dst = [DUT_MAC_ADDR]
+  src = [DUT_MAC_ADDR]
+  type = 0x800
+...
+```
+
+##### Get interfaces which are members of LAG, RIF and VLAN. Repeat defined test steps for each of those interfaces.
+
+##### Test steps
+- PTF host will send IP packet specifying identical src and dst MAC.
+- When packet reaches SONIC DUT, it should be dropped according to the test objective
+- Get L2 drop counter
+- Verify drop counter incremented
+- Get L3 drop counter
```

Comment:
In such case, test case should skip verification - "Verify L3 drop counter is not incremented".
It means that test case will only check that discard counter was incremented (without additional counters verification).
Information that vendor has one common counter can be stored in the same config file which defines CLI commands per vendor (described in section "Related DUT CLI commands").

So configuration example for vendor which has common drop counter can be like this:
For example:
```
- Vendor X:
    "get_l2_pkt_drops": "SHOW_CLI_CMD_FOR_COMBINED_DROPS"
    "get_l3_pkt_drops": "" # Not mandatory in this case
    "get_acl_pkt_drops": "" # Not mandatory in this case
    "parser": "path_to_the_vendor_x_parser.py"
    "combined_drop": True # test case will use 'get_l2_pkt_drops' to get all drops counter
...
```

Thoughts?

Reply:
> 
> 
> what if a vendor has combined l2/l3 drop counter?

In such case, test case should skip verification - "Verify L3 drop counter is not incremented".
It means that test case will only check that discard counter was incremented (without additional counters verification).
Information that vendor has one common counter can be stored in the same config file which defines CLI commands per vendor (described in section "Related DUT CLI commands").

So configuration example for vendor which has common drop counter can be like this:
For example:

- Vendor X:
    "get_l2_pkt_drops": "SHOW_CLI_CMD_FOR_COMBINED_DROPS"
    "get_l3_pkt_drops": "" # Not mandatory in this case
    "get_acl_pkt_drops": "" # Not mandatory in this case
    "parser": "path_to_the_vendor_x_parser.py"
    "combined_drop": True # test case will use 'get_l2_pkt_drops' to get all drops counter
...

Thoughts?

Reply:
I like this idea, but is the "combined_drop" field necessary? If get_l3_pkt_drops and get_acl_pkt_drops are optional anyways it seems a little redundant.

Reply:
Simplified approach. As different vendors can't have different SONiC CLI, CLI is common. However internal counters calculation can be different. So it is enough to know some vendor identifier to determine whether current DUT supports combined drop counters or not.
So platform name is selected for such reason.

Reply:
Is the idea here that any vendor on that list will simply use the results from portstat and any other vendor will use the three separate CLI commands?

Reply:
After more discussions I understand that customers can't have some specific commands to get counters as SONiC CLI should be common for any vendors. So here idea is that any vendor will use the results from "portstat", but counters calculation/representation for different vendors can be different.

Reply:
I understand that part, my question is about this file you added:
> tests/drop_counters/combined_drop_counters.yml
> ```
> - "[REGEXP FOR VENDOR X]"
> - "[REGEXP FOR VENDOR Y]"
> ```

Is this file a list of vendors with combined drop counters?

Reply:
Basically this is a list of a regular expressions for specific platforms, as usually platform name contains vendor name as well.

It will be better to rename it like:
- "[REGEXP FOR PLATFORM Y]"

Some explanation:
By the way framework will call "show platform summary" to get DUT platform name.
Then it will try to match obtained platform name to the defined regular expression.
If matched, test cases will do verification like with combined counters.

Reply:
Can you give an example of what the REGEXP would look like?

Reply:
For example if platform name is **x86_64-mlnx_msn2100-r0**
Regexp can be - **x86_64-mlnx**
Or you can use any other regexp to preciously detect expected platform name

Reply:
OK, I think I understand now. Sounds good. 👍